### PR TITLE
[WIP] Refactored Model

### PR DESCRIPTION
**Pull request by SidRoberts**
Don't merge this just yet - there'll be a few more commits in the next few days.

### CODE REVIEW

**Comment by sergeyklay** [comment]
@SidRoberts It is changes for `3.0.x` or for `3.1.x` ?

**Comment by SidRoberts** [comment]
3.0.x - I'm just simplifying the code in the existing methods for now.

**Comment by sergeyklay** [comment]
@SidRoberts It seems all ok with test. You can continue your work
Refs:
* https://github.com/Codeception/Specify/pull/35
* https://github.com/phalcon/cphalcon/pull/12332

File: phalcon/mvc/model/relation.zep [hunk 1]
```diff
@@ -157,14 +157,7 @@ class Relation implements RelationInterface
...
 	{
-		var options;
-		let options = this->_options;
-		if typeof options == "array" {
-			if isset options["foreignKey"] {
-				return true;
-			}
-		}
-		return false;
+		return isset this->_options["foreignKey"];
```
Comment (sergeyklay) [inline phalcon/mvc/model/relation.zep]:
```
protected _options;
```

not array

**Review by sergeyklay — CHANGES_REQUESTED** [review]

**Review by sergeyklay — APPROVED** [review]

File: phalcon/mvc/model/relation.zep (hunk 1 above)
Comment (Jurigag) [inline phalcon/mvc/model/relation.zep]: But won't php complain about this ?

Comment (Jurigag) [inline phalcon/mvc/model/relation.zep]: Yea i tested and it's fine in php.

**Comment by SidRoberts** [comment]
@sergeyklay Assuming this passes in Travis, I think it's ready to merge.

**Comment by Jurigag** [comment]
Squash please :)

**Comment by SidRoberts** [comment]
I'm not on my dev machine and Serghei can rebase it when merging.

**Comment by sergeyklay** [comment]
Thanks

**Comment by sergeyklay** [comment]
@SidRoberts Now builds for 3.0.x are broken. https://travis-ci.org/phalcon/incubator/builds/169510034

Could you please take a look

**Comment by SidRoberts** [comment]
Fixed in phalcon/incubator#680.

**Comment by Jurigag** [comment]
```
Warning: Variable "metaData" declared but not used in Phalcon\Mvc\Model\MetaData::writeMetaDataIndex in /root/cphalcon/phalcon/mvc/model/metadata.zep on 305 [unused-variable]

          var metaData, source, schema, key;

```

@SidRoberts

**Comment by sergeyklay** [comment]
Fixed

**Pull request merged**